include/crm/common/ipc_pacemakerd.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2020 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#ifndef PCMK__IPC_PACEMAKERD__H
+#  define PCMK__IPC_PACEMAKERD__H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \file
+ * \brief IPC commands for Pacemakerd
+ *
+ * \ingroup core
+ */
+
+#include <stdbool.h>                    // bool
+#include <libxml/tree.h>                // xmlNode
+#include <crm/common/ipc.h>             // pcmk_ipc_api_t
+
+enum pcmk_pacemakerd_state {
+    pcmk_pacemakerd_state_invalid = -1,
+    pcmk_pacemakerd_state_init = 0,
+    pcmk_pacemakerd_state_starting_daemons,
+    pcmk_pacemakerd_state_wait_for_ping,
+    pcmk_pacemakerd_state_running,
+    pcmk_pacemakerd_state_shutting_down,
+    pcmk_pacemakerd_state_shutdown_complete,
+    pcmk_pacemakerd_state_max = pcmk_pacemakerd_state_shutdown_complete,
+};
+
+//! Possible types of pacemakerd replies
+enum pcmk_pacemakerd_api_reply {
+    pcmk_pacemakerd_reply_unknown,
+    pcmk_pacemakerd_reply_ping,
+};
+
+/*!
+ * Pacemakerd reply passed to event callback
+ */
+typedef struct {
+    enum pcmk_pacemakerd_api_reply reply_type;
+
+    union {
+        // pcmk_pacemakerd_reply_ping
+        struct {
+            const char *sys_from;
+            enum pcmk_pacemakerd_state state;
+            time_t last_good;
+            int status;
+        } ping;
+    } data;
+} pcmk_pacemakerd_api_reply_t;
+
+int pcmk_pacemakerd_api_ping(pcmk_ipc_api_t *api, const char *ipc_name);
+unsigned int
+    pcmk_pacemakerd_api_replies_expected(pcmk_ipc_api_t *api);
```

Comment:
You don't need replies_expected for pacemakerd; the controller just needed that for compatibility with how crm_resource calls it since the controller API doesn't have the equivalent of call_id.

The idea is that the client has to match server replies to its requests somehow when it sends multiple requests. The call_id approach (cib, executor, fencer) is better; replies_expected is just a simple "N requests get N replies" counter.

I think the pacemakerd api is expected to remain simple enough (send one request, wait for one reply) that we don't need either approach. If we ever do, I'd rather the call_id approach. I'd love to eventually get to the point where call_id is baked into the generic model, but the need for backward api compatibility makes that a distant dream.

Reply:
Code was there, thought it might be useful for some debugging issue ... good advice to end up with a bloated API ;-)